### ISSUE TITLE
flaky e2e: Network when a minion node becomes unreachable [replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster

### ISSUE THREAD

**Issue by goltermann** [kubernetes/kubernetes#14695]
Nodes Network when a minion node becomes unreachable [replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster(Failed 17 times in the last 30 runs. Stability: 43 %)Failed with 0 retries

**Comment by brendandburns** [comment]
@davidopp for triage

**Comment by fgrzadkowski** [comment]
I'll look into this.

**Comment by davidopp** [comment]
I talked to @bprashanth about this and he had a pretty good guess in about two seconds about what the problem is.

ReplicationController sync method does this:
```
	if !rm.podStoreSynced() {
		// Sleep so we give the pod reflector goroutine a chance to run.
		time.Sleep(PodStoreSyncedPollPeriod)
		glog.Infof("Waiting for pods controller to sync, requeuing rc %v", rc.Name)
		rm.enqueueController(&rc)
		return nil
	}
```

but DaemonSet controller doesn't. (I guess when DaemonSet was copy-pasted from ReplicationController, this was accidentally removed.)

Pod reflector and syncer are running in separate goroutines, and if pod reflector doesn't run then syncer will think there are no pods on the node.

**Comment by davidopp** [comment]
To be clear: this is about the test
Controller Manager should not create/delete replicas across restart 
which is the one that is actually failing.

**Comment by fgrzadkowski** [comment]
I'm confused you're saying that this is about #14693, but you're updating #14695. I thought they are about two different tests...

**Comment by davidopp** [comment]
Yes sorry, I put my comment in the wrong issue. Reassigning this one to you, and I'll copy my comment into the other one.

**Comment by fgrzadkowski** [comment]
First failure:
```
Sep 29 18:00:37.181: Couldn't delete testing namespaces 'e2e-tests-resize-nodes-yd40e', Namespace e2e-tests-daemonsets-fecw4 is active
```
We check if all namespace are deleted. If other tests failed to delete their namespace this test will fail. I'm going to delete this logic as we already check if we have successfully deleted our namespace.

See also https://github.com/kubernetes/kubernetes/issues/13828#issuecomment-144748236 for some more details.

**Comment by fgrzadkowski** [comment]
Second failure:
```
00:50:13 Sep 29 00:50:13.154: INFO: Node e2e-flaky-minion-c1hs condition 1/1: type: Ready, status: Unknown, reason: "NodeStatusUnknown", message: "Kubelet stopped posting node status.", last transition time: 2015-09-29 00:44:12 -0700 PDT
00:50:13 Sep 29 00:50:13.154: INFO: Node e2e-flaky-minion-ryun condition 1/1: type: Ready, status: True, reason: "KubeletReady", message: "kubelet is posting ready status", last transition time: 2015-09-29 00:10:54 -0700 PDT
00:50:13 Sep 29 00:50:13.154: INFO: Successfully found node e2e-flaky-minion-ryun readiness to be true
00:50:13 
00:50:13 • Failure [415.780 seconds]
00:50:13 Nodes
00:50:13 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:553
00:50:13   Network
00:50:13   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:552
00:50:13     when a minion node becomes unreachable
00:50:13     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:551
00:50:13       [replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster [It]
00:50:13       /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:550
00:50:13 
00:50:13       Expected error:
00:50:13           <*errors.errorString | 0xc2080845c0>: {
00:50:13               s: "timed out waiting for the condition",
00:50:13           }
00:50:13           timed out waiting for the condition
00:50:13       not to have occurred
00:50:13 
00:50:13       /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:377
00:50:13 ------------------------------
```

We disable network by running the following command:
```
$ sudo iptables --insert OUTPUT --destination e2e-test-filipg-master --jump DROP
```
I checked and it seems that it results in ```curl```ing master to hang. I suspect that Kubelet hangs on updating node status and depending when it breaks the connection (by timeout) it will fit in test timeouts or not.

**Comment by fgrzadkowski** [comment]
@lavalamp What is the default timeout for connections using client libraries to apiserver? I checking the code, but can't find it.

**Comment by lavalamp** [comment]
I'm not sure if we set one, but we should. I guess I would expect to see
that in the config, w/ default set by this function:
https://github.com/kubernetes/kubernetes/blob/00ab055fa1798b1e9ad2b3535dd90dc021db3935/pkg/client/unversioned/helper.go#L288

**Comment by lavalamp** [comment]
So: let's add a timeout, and revert #14914.

**Comment by fgrzadkowski** [comment]
@lavalamp I think that adding a timeout and reverting #14914 are orthogonal. I we want to simulate network problems with node->master communication then using REJECT is better anyway. At the same time adding timeout makes sense for many other reasons.

**Comment by lavalamp** [comment]
OK then, let's add a timeout and test *both* DROP and REJECT-- no sort of network problem should hang a component.

**Comment by fgrzadkowski** [comment]
@lavalamp I think it didn't hang. It just that the kubelet was retrying too slow and sometimes it fit test timeout, sometime it didn't.

**Comment by lavalamp** [comment]
Could we reduce the timeout to make it always fit the test timeout?

On Tue, Oct 6, 2015 at 12:56 AM, Filip Grzadkowski <notifications@github.com

**Comment by fgrzadkowski** [comment]
@lavalamp I guess we can. I'll try tomorrow.

**Comment by fgrzadkowski** [comment]
It seems we have a new type of failure:
```
17:46:05 • Failure [650.683 seconds]
17:46:05 Nodes
17:46:05 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:539
17:46:05   Network
17:46:05   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:538
17:46:05     when a minion node becomes unreachable
17:46:05     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:537
17:46:05       [replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster [It]
17:46:05       /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:536
17:46:05 
17:46:05       Expected error:
17:46:05           <*errors.errorString | 0xc2084fc520>: {
17:46:05               s: "failed to wait for pods running: [gave up waiting for pod 'additionalpod' to be 'running' after 5m0s]",
17:46:05           }
17:46:05           failed to wait for pods running: [gave up waiting for pod 'additionalpod' to be 'running' after 5m0s]
17:46:05       not to have occurred
17:46:05 
17:46:05       /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:526
```
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-flaky/5644/consoleFull

It seems that kubelet on a machine where network is blocked doesn't notice new pod. After reading logs I assume that kubelet doesn't notice that the watch is broken and doesn't try to relist everything. It notices new pod only after watch timeout.

**Comment by lavalamp** [comment]
Ugh, I hope we don't have to start sending watch heartbeats.

**Comment by wojtek-t** [comment]
It seems that all recent failures are with the following error:

    Expected error:
    01:24:07           <*errors.errorString | 0xc208126130>: {
    01:24:07               s: "timed out waiting for the condition",
    01:24:07           }
    01:24:07           timed out waiting for the condition
    01:24:07       not to have occurred
    01:24:07 
    01:24:07       /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:380

which means the reason is here:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/resize_nodes.go#L380

@fgrzadkowski - can you please take a look or delegate?

**Comment by wojtek-t** [comment]
I did some more debugging of it.

I was looking into this failure:
http://kubekins.dls.corp.google.com/view/Flaky/job/kubernetes-e2e-gce-flaky/8132/consoleFull

There is something strange happening here:

1. We run performTemporaryNetworkFailure:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/resize_nodes.go#L295
2. We succesfully block the load from kubelet:
3. And the succesfully wait for node to become not ready:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/resize_nodes.go#L374

        01:18:05 Nov 30 01:18:05.820: INFO: Successfully found condition Ready of node e2e-flaky-minion-fsxc to be false

4. After some time nodecontroller successfully removes the pod:

        // controller logs:
        I1130 09:22:25.381404       6 nodecontroller.go:419] Evicting pods on node e2e-flaky-minion-fsxc: 2015-11-30 09:22:25.381392699 +0000 UTC is later than 2015-11-30 09:18:05.116010276 +0000 UTC + 4m20s
        I1130 09:22:25.405692       6 nodecontroller.go:739] Starting deletion of pod my-hostname-net-2vhn6
 
        // apiserver logs:   
        I1130 09:22:25.413139       7 handlers.go:130] DELETE /api/v1/namespaces/e2e-tests-resize-nodes-zg3tn/pods/my-hostname-net-2vhn6: (6.655367ms) 200 [[kube-controller-manager/v1.2.0 (linux/amd64) kubernetes/c1af9dc] 127.0.0.1:51669]

5. Then replication controller creates a new pod:

        // controller logs:
        I1130 09:22:25.421267       6 replication_controller.go:356] Too few "e2e-tests-resize-nodes-zg3tn"/"my-hostname-net" replicas, need 3, creating 1

        // apiserver logs:
        I1130 09:22:25.433384       7 handlers.go:130] POST /api/v1/namespaces/e2e-tests-resize-nodes-zg3tn/pods: (10.407029ms) 201 [[kube-controller-manager/v1.2.0 (linux/amd64) kubernetes/c1af9dc] 127.0.0.1:51676]

        // events printed in test:
        01:23:06 Nov 30 01:23:06.192: INFO: event for my-hostname-net: {replication-controller } SuccessfulCreate: Created pod: my-hostname-net-y91vd
        01:23:06 Nov 30 01:23:06.192: INFO: event for my-hostname-net: {replication-controller } SuccessfulCreate: Created pod: my-hostname-net-2vhn6
        01:23:06 Nov 30 01:23:06.192: INFO: event for my-hostname-net: {replication-controller } SuccessfulCreate: Created pod: my-hostname-net-nqjiu
        01:23:06 Nov 30 01:23:06.192: INFO: event for my-hostname-net: {replication-controller } SuccessfulCreate: Created pod: my-hostname-net-5js2f

6. And here is something I don't undertand - because of some reason, the test doesn't observe that the pod actually disappeared.

@gmarek @davidopp - any thoughts on it?

**Comment by wojtek-t** [comment]
BTW - this is exactly the same situation in:
http://kubekins.dls.corp.google.com/view/Flaky/job/kubernetes-e2e-gce-flaky/8129/consoleFull

**Comment by wojtek-t** [comment]
OK - I think this is related to default 30s deletion grace period.

**Comment by wojtek-t** [comment]
Yes - it seems that this is exactly the case.

I will send out PR fixing this out for review.

**Comment by wojtek-t** [comment]
I'm reopening it for tracking.
If there won't be any failures within next few days - I will move the test back from flaky and close this issue.

**Comment by davidopp** [comment]
Thanks for fixing, @wojtek-t .

**Comment by wojtek-t** [comment]
It seems that the main issue is fixed.

However, there was one failure in the last 30 runs:
http://kubekins.dls.corp.google.com/view/Flaky/job/kubernetes-e2e-gce-flaky/8153/console

in which one pod simply didn't started (and it was "preparation" for the test, not the test itself).

**Comment by wojtek-t** [comment]
During next 2 days there was exactly one more failure of this test:
http://kubekins.dls.corp.google.com/view/Flaky/job/kubernetes-e2e-gce-flaky/8225/console

it is exactly the same as the previous one.
However, those are duplicates of https://github.com/kubernetes/kubernetes/issues/17931

So I'm going to move this test out of flaky suite.
